### PR TITLE
refactor: add flexibility to plot_upgrades_map

### LINKED ISSUE #267: Transmission upgrades visualization refactor
As discussed with @danlivengood , we need to refactor [map_upgrades](https://github.com/Breakthrough-Energy/PostREISE/blob/develop/postreise/plot/plot_upgrades_map.py#L259) function in `plot_upgrades_map.py` to make it compatible for individual interconnections (now it is hard coded for USA scenarios only).

### PR DESCRIPTION
**Pull request by danielolsen**
[Pull Request doc](https://breakthrough-energy.github.io/docs/user/git_guide.html#d-pull-request)

### Purpose
Allow transmission upgrades plots to be created using any interconnection, and to properly plot new branches as well as scaled branches.
Closes #267.

### What the code is doing
In **upgrades.py**: we fix the merging of the two branch data frames to ensure that the lat/lon columns of the merged data frame include the lat/lon data of new branches.

In **plot_upgrades_map.py**:
- We make use of the `plot_states` function to simplify border plotting.
- Previously-hardcoded b2b_indices are now passed by the user.
- Previously-hardcoded figure sizes can now be passed by the user (default sizing remains, which is good for plotting USA).
- The user can specify explicit `x_range` and `y_range`, to be able to zoom into certain regions.
- Scaling factors are now specified in pixels/GW, not pixels/MW, based on conversations in https://github.com/Breakthrough-Energy/PostREISE/pull/266.

### Testing
@danlivengood has been using this code to plot maps of Eastern scenarios with new branches.

### Time estimate
10 minutes. The data frame merge is the most complex part, most everything else is just adding flexibility to previously-hardcoded values.

### CODE REVIEW

File: postreise/plot/plot_upgrades_map.py [hunk 1]
```diff
@@ -81,7 +82,7 @@ def _map_upgrades(
         {
             "xs": branch_all[["from_x", "to_x"]].values.tolist(),
             "ys": branch_all[["from_y", "to_y"]].values.tolist(),
-            "cap": branch_all["rateA"] * all_branch_scale + all_branch_min,
+            "cap": branch_all["rateA"] / 1000 * all_branch_scale + all_branch_min,
```
Comment (jon-hagg) [inline postreise/plot/plot_upgrades_map.py:85]: Consider making the 1000 a module constant

Comment (danielolsen) [reply postreise/plot/plot_upgrades_map.py:85]: I'm trying to come up with a reasonable name, but they all seem a little silly, e.g. `num_MW_in_GW`. What about adding a new column `branch_all["rateA_GW"] = branch_all["rateA"] / 1000  # MW to GW` and then using that column later in the function?

File: postreise/analyze/transmission/upgrades.py [hunk 2]
```diff
@@ -35,6 +35,12 @@ def get_branch_differences(branch1, branch2):
         branch2, how="outer", right_index=True, left_index=True, suffixes=(None, "_2")
     )
     branch_merge["diff"] = branch_merge.rateA_2.fillna(0) - branch_merge.rateA.fillna(0)
+    # Ensure that lats & lons get filled in as necessary from branch2 entries
```
[the remark this answers is not in the document]
Comment (danielolsen) [reply postreise/analyze/transmission/upgrades.py:38]: `dataframe` -> `data frame`?

File: postreise/plot/plot_upgrades_map.py (hunk 1 above)
Comment (jon-hagg) [reply postreise/plot/plot_upgrades_map.py:85]: The new column would make sense but we use the same factor in several places, so was hoping to reuse that. What about calling it `scale_factor` and just putting a comment next to it?

File: postreise/analyze/transmission/upgrades.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jon-hagg) [reply postreise/analyze/transmission/upgrades.py:38]: Heh sorry, was looking at "branch1 and branch2 both much have 'rateA' columns"

[the remark this answers is not in the document]
Comment (danielolsen) [reply postreise/analyze/transmission/upgrades.py:38]: Oh I see, `much have` -> `must have`. Fixed.

**Review by BainanXia — APPROVED** [review]
The logic makes sense. Given Research Dan has been testing this for multiple times, I believe this is good to go.

**Review by jon-hagg — APPROVED** [review]

File: postreise/plot/plot_upgrades_map.py [hunk 3]
```diff
@@ -7,41 +7,42 @@
...
-    all_branch_scale=1e-3,
-    diff_branch_scale=1e-3,
+    all_branch_scale=1,
+    diff_branch_scale=0.5,
     all_branch_min=0.1,
     diff_branch_min=1.0,
     b2b_scale=5e-3,
+    figsize=(1400, 800),
+    x_range=None,
+    y_range=None,
```
Comment (rouille) [inline postreise/plot/plot_upgrades_map.py:26]: What about `background_map` to display or not the tile

File: postreise/plot/plot_upgrades_map.py [hunk 4]
```diff
@@ -7,41 +7,42 @@
...
-    b2b_indices,
+    b2b_indices=None,
     diff_threshold=100,
-    all_branch_scale=1e-3,
-    diff_branch_scale=1e-3,
+    all_branch_scale=1,
+    diff_branch_scale=0.5,
     all_branch_min=0.1,
     diff_branch_min=1.0,
     b2b_scale=5e-3,
```
Comment (rouille) [inline postreise/plot/plot_upgrades_map.py:23]: Is that possible to have the factor around 1 (as we discussed previously)

File: postreise/plot/plot_upgrades_map.py [hunk 5]
```diff
@@ -129,10 +133,12 @@ def _map_upgrades(
...
         x_axis_location=None,
         y_axis_location=None,
-        plot_width=1400,
-        plot_height=800,
+        plot_width=figsize[0],
+        plot_height=figsize[1],
         output_backend="webgl",
         match_aspect=True,
+        x_range=x_range,
+        y_range=y_range,
```
[the remark this answers is not in the document]
Comment (danielolsen) [reply postreise/plot/plot_upgrades_map.py:145]: Done.

File: postreise/plot/plot_upgrades_map.py (hunk 3 above)
Comment (danielolsen) [reply postreise/plot/plot_upgrades_map.py:26]: I added the broader `plot_states_kwargs` logic, covering `background_map` and other things.

File: postreise/plot/plot_upgrades_map.py (hunk 4 above)
Comment (danielolsen) [reply postreise/plot/plot_upgrades_map.py:23]: They will be 1 by default (1 pixel/GW).

File: postreise/plot/plot_upgrades_map.py (hunk 1 above)
Comment (danielolsen) [reply postreise/plot/plot_upgrades_map.py:88]: `scale_factor` seems pretty generic. What if we create new variables `all_branch_scale_MW` and `diff_branch_scale_MW`, calculate them at the top with the `/ 1000` logic and a comment, and then apply them everywhere?

File: postreise/plot/plot_upgrades_map.py (hunk 4 above)
Comment (rouille) [reply postreise/plot/plot_upgrades_map.py:23]: Not sure I understand. The default value is `5e-3`

Comment (danielolsen) [reply postreise/plot/plot_upgrades_map.py:23]: Ah, I see which line you're referring to now. Yes, I'll make that change. The new factor will be `5`.

Comment (danielolsen) [reply postreise/plot/plot_upgrades_map.py]: Done.

File: postreise/plot/plot_upgrades_map.py (hunk 1 above)
Comment (jon-hagg) [reply postreise/plot/plot_upgrades_map.py:88]: Sounds good to me

Comment (danielolsen) [reply postreise/plot/plot_upgrades_map.py]: Added in the latest commit.
https://github.com/Breakthrough-Energy/PostREISE/pull/271/commits/b1b7bc1d10dbf78d989f96dd76b5cb9557e9524f

Comment (jon-hagg) [reply postreise/plot/plot_upgrades_map.py]: Thanks for indulging

**Review by rouille — APPROVED** [review]
Thanks

**Pull request merged**